File: 1857_Largest_Color_Value_in_a_Directed_Graph.js
/**
 * @param {string} colors
 * @param {number[][]} edges
 * @return {number}
 */
var largestPathValue = function(colors, edges) {
    const n = colors.length;
    const graph = Array.from({ length: n}, () => []);
    const inDegree = new Array(n).fill(0);
    const dp = Array.from({ length: n}, () => new Array(26).fill(0));

    for(const [u, v] of edges) {
        graph[u].push(v);
        inDegree[v]++;
    }
    let queue = []
    for(let i = 0; i < n; i++) {
        if(inDegree[i] === 0) {
            queue.push(i);
        }
    }

    let processed = 0, maxColorValue = 0;

    while(queue.length) {
        let node = queue.shift();
        processed++;
        let colorIndex = colors.charCodeAt(node) - 97;
        dp[node][colorIndex]++;
        maxColorValue = Math.max(maxColorValue, dp[node][colorIndex]);

        for(let neighbor of graph[node]) {
            for(let i = 0; i < 26; i++) {
                dp[neighbor][i] = Math.max(dp[neighbor][i], dp[node][i]);
            }
            if(--inDegree[neighbor] == 0) {
                queue.push(neighbor);
            }
        }
    }

    return processed === n ? maxColorValue : -1;
};

File: 1857_Largest_Color_Value_in_a_Directed_Graph.rs
use std::collections::{VecDeque, HashMap};

impl Solution {
    pub fn largest_path_value(colors: String, edges: Vec<Vec<i32>>) -> i32 {
        let n = colors.len();
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n];
        let mut in_degrees = vec![0; n];
        let mut queue = VecDeque::new();
        let mut count = vec![vec![0; 26]; n];

        // Build the graph.
        for edge in edges.iter() {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            graph[u].push(v);
            in_degrees[v] += 1;
        }

        // Perform topological sorting.
        for i in 0..n {
            if in_degrees[i] == 0 {
                queue.push_back(i);
            }
        }

        let mut processed = 0;
        let mut ans = 0;

        while let Some(out) = queue.pop_front() {
            processed += 1;
            let color_idx = colors.as_bytes()[out] as usize - b'a' as usize;
            count[out][color_idx] += 1;
            ans = ans.max(count[out][color_idx]);

            for &in_node in &graph[out] {
                for i in 0..26 {
                    count[in_node][i] = count[in_node][i].max(count[out][i]);
                }
                in_degrees[in_node] -= 1;
                if in_degrees[in_node] == 0 {
                    queue.push_back(in_node);
                }
            }
        }

        if processed == n {
            ans
        } else {
            -1
        }
    }
}
